### ISSUE TITLE
Umbrella: Adding in functionality to create an ObsPack

### ISSUE THREAD

**Issue by rt17603** [openghg/openghg#1078]
We want to be able to generate an ObsPack from an openghg object store. Some of the specifics we may want to split out into a separate or multiple separate repositories but for now we shall collect together the steps here. 

Steps:
- Define what goes into an obspack
   - Set up a very explicit config file which can be used to create search queries (could be a csv, json or toml)
   - Set up a special keyword "tag" which can be used to collect together data to include in an obspack
- Side check: ensure output is suitably standardised 
   - #1077 - links to making sure this is done for ICOS
- Search object store and retrieve the data (based on config initially and them generalising to be able to use tags)
- Collect together this data and output to the obspack framework
   - Need details to include in obspack framework and defined structure
   - Use metadata to create filenames for output nc files

**Comment by rt17603** [comment]
We discussed recording actual UUIDs, versions for the data and search metadata to make sure creation of a obspack from a given object store is reproducible. This will be important once we have the fundamentals of this down and as we start to improve the process.

**Comment by rt17603** [comment]
Suggestion from @brendan-m-murphy 

> You could parallelise this easily with dask "bags" if you like. Maybe after validating file names, make a dask "bag" containing tuples: (ds, output_name) and then apply a function like lambda x: x[0].to_netcdf(x[1]). If the output files are all unique, then you can just write everything in parallel.
>
> (If you do this, you would probably want to use a different loop to make the site_details_df.)
>
> (Also, it might be better to make the bag contain tuples (obs_data, output_name) and use .map with lambda x: x[0].data.to_netcdf(x[1]), since calling .data might trigger something that dask would like to know about.)

**Comment by rt17603** [comment]
Excellent restructure suggestion based on current PR #1117 

> There's a lot of code here. I think it might be easier to understand if the code modelled the ObsPack more closely.
> 
> The code could broadly split into two phases:
> 
>  -  parse the search input to make the obspack structure (which would be represented in an ObsPack object)
>  - write out the data
> 
> Obviously this is what the code is already doing, but the point would be to focus on building the obs pack structure in the code using objects. Right now it feels like the structure is encoded in many functions that pass file paths around.
> 
> In addition to your StoredData object, you could have a Collection object (in the comments, I suggested ObsPack and Subfolder objects, although they both share some common features) that contains StoreData and other Collection objects. Basically, the Collection object is like a folder, and the StoredData object is like a file.
>
> Also, maybe some of the code can be split into separate modules? For instance, a submodule around the code that parses the input files and organises the search results.
> This might also include the code that builds the ObsPack, Subfolder, StoredData objects.
>
> An ObsPack object could also bring together some of the code into a logical collection, etc.
>
> I'm happy to try to implement some of these suggestions if you don't have time.
>
> Anyway, overall, there is lots of good functionality, but continuing what you started with the StoredData objects would bring some nice structure to the code.

**Comment by rt17603** [comment]
Additional useful feedback on this:

> It seems like the StoredData object is constructing all of the path info itself.
> 
> Maybe create ObsPack and Subfolder objects to share the load? These could have a path attribute, and a method write(self, path) that write the data to path/self.path. (The default value for path could be Path.cwd().)
> 
> ObsPack and Subfolder objects could store more Subfolder objects, as well as StoredData objects.
> 
> So e.g. ObsPack could have a collection of these things ObsPack.data : list[Subfolder | StoredData], and its write method could be like this:
> 
>     def write(self, path: str | Path | None = None):
>       if path is None:
>           path = self.path
>       else:
>           path = Path(path) / self.path
>       path.mkdir(exists_ok=True, parents_ok=True)  # I forget the actual names of the flags
> 
>     for x in self.data:
>         x.write(path)
> 
> The Subfolder.write method would be the same, and the StoredData.write method would actually write the data (and would construct its filename, then add that to the end of the path passed to write).
> 
> The ObsPack and Subfolder objects could have an add method (or maybe add_subfolder, add_stored_data) that could pass along info that is the same for all objects in that folder. This method might also be able to check if the new StoredData object created would have the same name as one that was already added.

**Comment by rt17603** [comment]
Suggestion for the file naming creation function:

> You could split this up into a function that makes template strings, and then you can fill the values using `str.format()`. So the default case would be to return `f"{{{keys}}}"` (or `"{" + keys + "}"`). Then you call `template_string.format(**dictionary)` https://docs.python.org/3/library/string.html#format-string-syntax
> 
> A draft version is:
> ```
> def _construct_name_template(keys: str | list, separators: Sequence[str]) -> str:
>     if isinstance(keys, list):
>         sep, *separators = separators
>         values = (_construct_name_template(key, separators) for key in keys)
>         return sep.join(values)
>     return f"{{{keys}}}"
> ```

> for example:

> `>> template = _construct_name_template([["species", "sector"], "site", "inlet"], ["_", "-"])`
> `>> template`
'{species}-{sector}_{site}_{inlet}'

**Comment by rt17603** [comment]
TODO (next stage after #1117 is ready)
- Make better use of the `StoredData` object and think about higher level structure with an `ObsPack` object. [larger - may need an issue]
    - Think about this in the context of a more general DataPack as well
- Resolve issue with versions also playing into the uniqueness check
   - ** May want to check the values related to name_components for this or formalise the filename_stub idea
- Allow name_components top level dict input to populate itself with defaults for missing obs_types as necessary

Questions: 
- Would it be more useful to allow different inputs to create_obspack?
  - e.g. a DataFrame equivalent of the csv file (kind of already included but not at top create_obspack layer yet)
  - e.g. a list of SearchResults?
  - e.g. a dictionary of keys rather than a DataFrame
- Allow inclusion of multiple search results e.g. this could be the right approach for including the tag?
  - The get_functions would reject this at the moment..
  - We need get because of the inlet combination but we potentially need search because of the ability to extract multiple datasets
- Could we make choosing of surface versus column smarter - can we tie into platform better?
- Do we want to get rid of pre-creation of the obspack structure and allow this to be created as the files are added instead?
  - Could end up with misalignment otherwise

**Comment by rt17603** [comment]
Example of current "obspack" that Joe creates can be found on BP1: /group/chem/acrg/obs_frozen/2025-07-04

Should make sure an obspack with looks exactly like this can be created using this framework.

**Comment by ben-westhenry** [comment]
After quick Ben and Rachel catch up. Ben to:

-Sort ICOS authentication  (see if this resolves local installation errors.)
-Setup openghg on blue pebble (so can access data.)
-Add path to config file to read in object store (data) for GEMMA obspack
-See if can run Gemma obspack when reading in correct object store.
-Look at obspack code and think of wider obspack cases (wider than purely observation data) e.g footprints etc, and scope how the pipeline might look (i.e adding this additional functionality with minimum number of tweaks to the current code and what steps that would involve).
-Can we create an obspack  that’s like GEMMA obspack pipeline that contains same structure Joe wants? Example of current "obspack" that Joe creates can be found on BP1: /group/chem/acrg/obs_frozen/2025-07-04. Should make sure an obspack which looks exactly like this can be created using this framework.
(-When ICOS SPARQL interface done ask Joe to test and see if he is happy with it?)
